File: be/src/main/java/sg/toss_sg/configs/DatabaseConnectionPool.java
package sg.toss_sg.configs;

import java.sql.Connection;
import java.sql.SQLException;

import javax.sql.DataSource;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;

@Configuration
public class DatabaseConnectionPool {

    private static DatabaseConnectionPool databaseConnectionPool = new DatabaseConnectionPool();
    private DataSource dataSource;

    private DatabaseConnectionPool() {
        HikariConfig config = new HikariConfig();

        // Database settings
        config.setJdbcUrl("jdbc:postgresql://localhost:5432/zen_dev"); // Database URL
        config.setUsername("zen_developer"); // Database username
        config.setPassword("airbusa380861"); // Database password

        // HikariCP settings
        config.setMaximumPoolSize(100); // Maximum number of connections
        config.setMinimumIdle(2); // Minimum number of idle connections
        config.setIdleTimeout(30000); // Idle timeout in ms
        config.setMaxLifetime(1800000); // Max lifetime of a connection in ms
        config.setConnectionTimeout(20000); // Connection timeout in ms

        // Create the DataSource

        this.dataSource = new HikariDataSource(config);
        DatabaseConnectionPool.databaseConnectionPool = this;
    }

    public static DatabaseConnectionPool getInstance() {
        return databaseConnectionPool;
    }

    @Bean
    public static DatabaseConnectionPool databaseConnectionPool() {
        return databaseConnectionPool;
    }

    public Connection getConnection() {
        try {
            return this.dataSource.getConnection();
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: be/src/main/kotlin/sg/flow/configs/DatabaseConnectionPool.kt
package sg.flow.configs

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import java.sql.Connection
import java.sql.SQLException
import javax.sql.DataSource
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class DatabaseConnectionPool private constructor() {

    private val dataSource: DataSource

    init {
        val config =
                HikariConfig().apply {
                    jdbcUrl = "jdbc:postgresql://localhost:5432/flow_dev"
                    username = "flow_developer"
                    password = "flowdev123"

                    // HikariCP settings
                    maximumPoolSize = 100
                    minimumIdle = 2
                    idleTimeout = 30_000
                    maxLifetime = 1_800_000
                    connectionTimeout = 20_000
                }

        dataSource = HikariDataSource(config)
        instance = this
    }

    fun getConnection(): Connection? =
            try {
                dataSource.connection
            } catch (e: SQLException) {
                e.printStackTrace()
                null
            }

    companion object {
        @Volatile private var instance: DatabaseConnectionPool? = null

        @JvmStatic
        fun getInstance(): DatabaseConnectionPool =
                instance
                        ?: synchronized(this) {
                            instance ?: DatabaseConnectionPool().also { instance = it }
                        }

        @Bean @JvmStatic fun databaseConnectionPool(): DatabaseConnectionPool = getInstance()
    }
}
